### PR TITLE
x86: Preserve xmm4 in fpRem helper

### LINKED ISSUE #17129: JDK 8/11/17/18 - wrong result from JIT compilation with modular operation and field access
# JDK 8/11/17/18 - wrong result from JIT compilation with modular operation and field access

### Java version
$ java -version
```
openjdk 17.0.6 2023-01-17
IBM Semeru Runtime Open Edition 17.0.6.0 (build 17.0.6+10)
Eclipse OpenJ9 VM 17.0.6.0 (build openj9-0.36.0, JRE 17 Linux amd64-64-Bit Compressed References 20230117_397 (JIT enabled, AOT enabled)
OpenJ9   - e68fb241f
OMR      - f491bbf6f
JCL      - 927b34f84c8 based on jdk-17.0.6+10)
```

### Operating system details
$ cat /etc/*release
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.3 LTS"
NAME="Ubuntu"
VERSION="20.04.3 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.3 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

$ uname -a
```
Linux luzhou 5.4.0-86-generic #97-Ubuntu SMP Fri Sep 17 19:19:40 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

## Description

Incorrect result is given when running the following program. The bug affects 8u362, 11.0.18.0, 17.0.6.0, 18.0.2.1.

## Steps to reproduce
The following steps shows how to reproduce the bug on JDK 17.0.6.0 in a Ubuntu Linux environment.

### Compile
$ javac C.java

### Execute
$ java C
13815.0 // nondeterministic value

$ java -Xjit:optlevel=cold C 
17180.0 // nondeterministic value

$ java -Xnojit C
100000.0

### Source code for an executable test case
// C.java
```
public class C {

    private double q0;
    private double q1;
    private double q2;
    private double q3;

    public C(double a0, double a1, double a2, double a3) {
        this.q0 = a3;
        this.q1 = a1;
        this.q2 = 0;
        this.q3 = 0;
    }

    public static double m(double d) {
        C c = new C(0, 1.0, 0, d % d);
        return c.q1;
    }

    public static void main(String[] args) {
        double sum = 0;
        for (int i = 0; i < 100_000; ++i) {
            // m(1.0) is expected to be 1.0, but after a few thousand
            // iterations, m(1.0) returns 0.0.
            sum += m(1.0);
        }
        System.out.println(sum); // Expect 100000.0
    }
}
```

### Workaround
Disable JIT.
$ java -Xnojit C

### PR DESCRIPTION
**Pull request by BradleyWood**
Fixes #17129

### CODE REVIEW

**Comment by BradleyWood** [comment]
@hzongaro Would you mind reviewing?

**Comment by hzongaro** [comment]
Brad @BradleyWood, I haven't ever needed to look in detail at the various linkage conventions.  Can you point me to the description of the linkage convention that would apply for this helper?  I just wanted to understand why the helper should be responsible for preserving `xmm4` rather than the caller.

**Comment by hzongaro** [comment]
In an off-line discussion, Daryl @0xdaryl indicated that the callee should preserve that register, so I'll proceed with the review.

File: runtime/compiler/x/runtime/X86Codert.nasm [hunk 1]
```diff
@@ -487,8 +487,9 @@ _dblRemain:
         push _rax
         push _rbx
         push _rcx
-        sub _rsp, 48                              ; _rsp = _rsp - 48(bytes)
+        sub _rsp, 56                              ; _rsp = _rsp - 56(bytes)
```
Comment (hzongaro) [inline runtime/compiler/x/runtime/X86Codert.nasm:490]: Sorry for the basic question, but is it necessary to change the size here, and in other places in this function, from 48 to 56 bytes?  It looks like nothing is ever stored from offset 48 upwards, and xmm4 itself is stored at offset 24.

If this really is necessary, then I believe the [code under the `LARGE_NUMS:` case](https://github.com/eclipse-openj9/openj9/blob/aaa589e463a92685205b4cd330e59d4cdef22032/runtime/compiler/x/runtime/X86Codert.nasm#L626-L639) below would also need to add 56 to rsp on return.

Comment (BradleyWood) [reply runtime/compiler/x/runtime/X86Codert.nasm:490]: Yes, you are right

**Review by hzongaro — APPROVED** [review]
Looks good.  Thanks!

**Comment by hzongaro** [comment]
Jenkins test sanity xlinux,win,osx jdk11,jdk17

**Comment by hzongaro** [comment]
Testing has completed successfully.  Merging.

**Pull request merged**